File: app/src/main/java/com/osfans/trime/ime/symbol/SimpleAdapter.java
package com.osfans.trime.ime.symbol;

import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.blankj.utilcode.util.SizeUtils;
import com.osfans.trime.data.theme.FontManager;
import com.osfans.trime.data.theme.Theme;
import com.osfans.trime.databinding.SimpleItemOneBinding;
import com.osfans.trime.databinding.SimpleItemRowBinding;
import java.util.ArrayList;
import java.util.List;

public class SimpleAdapter extends RecyclerView.Adapter<SimpleAdapter.ViewHolder> {
  private final @NonNull List<SimpleKeyBean> mBeans = new ArrayList<>();
  private final @NonNull List<List<SimpleKeyBean>> mBeansByRows = new ArrayList<>();

  @NonNull
  public final List<SimpleKeyBean> getBeans() {
    return mBeans;
  }

  private final Theme theme;
  private final int columnSize;

  public SimpleAdapter(@NonNull Theme theme, int columnSize) {
    this.theme = theme;
    this.columnSize = columnSize;
  }

  public void updateBeans(List<SimpleKeyBean> beans) {
    mBeans.clear();
    mBeans.addAll(beans);
    mBeansByRows.clear();

    ArrayList<SimpleKeyBean> t = new ArrayList<>();
    for (int i = 0; i < mBeans.size(); i++) {
      t.add(mBeans.get(i));
      if ((i + 1) % columnSize == 0) {
        mBeansByRows.add(t);
        t = new ArrayList<>();
      }
    }
    if (!t.isEmpty()) {
      mBeansByRows.add(t);
    }

    notifyDataSetChanged();
  }

  @Override
  public int getItemCount() {
    return mBeansByRows.size();
  }

  @Override
  public long getItemId(int position) {
    return position * 1000L;
  }

  @NonNull
  @Override
  public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    final SimpleItemRowBinding binding =
        SimpleItemRowBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
    int size = SizeUtils.dp2px(theme.liquid.getFloat("single_width"));
    ViewGroup.LayoutParams p = new ViewGroup.LayoutParams(size, size);

    List<SimpleItemOneBinding> bindings = new ArrayList<>();
    for (int i = 0; i < columnSize; i++) {
      SimpleItemOneBinding view =
          SimpleItemOneBinding.inflate(LayoutInflater.from(parent.getContext()), null, false);
      bindings.add(view);
      binding.wrapper.addView(view.getRoot(), p);
    }
    ViewHolder holder = new ViewHolder(binding, bindings);

    for (int i = 0; i < holder.simpleKeyTexts.size(); i++) {
      holder.wrappers.get(i).setTag((Integer) i);

      TextView textView = holder.simpleKeyTexts.get(i);

      textView.setTextColor(theme.colors.getColor("key_text_color"));
      textView.setTypeface(FontManager.getTypeface(theme.style.getString("key_font")));
      textView.setGravity(Gravity.CENTER);
      textView.setEllipsize(TextUtils.TruncateAt.MARQUEE);

      final float labelTextSize = theme.style.getFloat("label_text_size");
      if (labelTextSize > 0) textView.setTextSize(labelTextSize);

      textView.setBackground(
          theme.colors.getDrawable(
              "key_back_color", "key_border", "key_border_color", "round_corner", null));
    }
    return holder;
  }

  public static class ViewHolder extends RecyclerView.ViewHolder {
    List<TextView> simpleKeyTexts;
    List<ViewGroup> wrappers;

    public ViewHolder(@NonNull SimpleItemRowBinding binding, List<SimpleItemOneBinding> views) {
      super(binding.getRoot());

      simpleKeyTexts = new ArrayList<>();
      wrappers = new ArrayList<>();
      for (int i = 0; i < views.size(); i++) {
        simpleKeyTexts.add((TextView) (views.get(i).getRoot().getChildAt(0)));
        ((ViewGroup) views.get(i).getRoot()).getChildAt(1).setVisibility(View.GONE);
        wrappers.add(views.get(i).getRoot());
      }
    }
  }

  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
    final List<SimpleKeyBean> bean = mBeansByRows.get(position);

    for (int i = 0; i < holder.simpleKeyTexts.size(); i++) {
      holder.simpleKeyTexts.get(i).setText("");
      if (i < bean.size()) {
        holder.wrappers.get(i).setVisibility(View.VISIBLE);
        holder.simpleKeyTexts.get(i).setText(bean.get(i).getLabel());
      } else {
        holder.wrappers.get(i).setVisibility(View.INVISIBLE);
      }

      if (listener != null) {
        holder
            .wrappers
            .get(i)
            .setOnClickListener(
                view -> {
                  if (view.getTag() != null) {
                    listener.onSimpleKeyClick(position * columnSize + (int) view.getTag());
                  }
                });
      }
    }
  }

  /*=====================添加OnItemClickListener回调================================*/
  public interface Listener {
    void onSimpleKeyClick(int position);
  }

  private Listener listener;

  public void setListener(@NonNull Listener listener) {
    this.listener = listener;
  }
}

File: app/src/main/java/com/osfans/trime/ime/symbol/SimpleAdapter.kt
// SPDX-FileCopyrightText: 2015 - 2024 Rime community
//
// SPDX-License-Identifier: GPL-3.0-or-later

package com.osfans.trime.ime.symbol

import android.text.TextUtils
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.osfans.trime.data.theme.ColorManager
import com.osfans.trime.data.theme.FontManager
import com.osfans.trime.data.theme.Theme
import com.osfans.trime.databinding.SimpleItemOneBinding
import com.osfans.trime.databinding.SimpleItemRowBinding
import splitties.dimensions.dp

class SimpleAdapter(
    private val theme: Theme,
    private val columnSize: Int,
) : RecyclerView.Adapter<SimpleAdapter.ViewHolder>() {
    private val mBeans = mutableListOf<SimpleKeyBean>()
    private val mBeansByRows = mutableListOf<List<SimpleKeyBean>>()

    fun updateBeans(beans: List<SimpleKeyBean>) {
        val prevSize = mBeansByRows.size
        mBeans.clear()
        notifyItemRangeRemoved(0, prevSize)
        mBeans.addAll(beans)
        mBeansByRows.clear()
        mBeansByRows.addAll(beans.chunked(columnSize))
        notifyItemRangeInserted(0, mBeansByRows.size)
    }

    override fun getItemCount(): Int = mBeansByRows.size

    override fun getItemId(position: Int): Long = position * 1000L

    private val mSingleWidth = theme.liquidKeyboard.singleWidth
    private val mSingleHeight = theme.liquidKeyboard.keyHeight
    private val mStringMarginX = theme.liquidKeyboard.marginX
    private val mTextSize = theme.generalStyle.labelTextSize
    private val mTextColor = ColorManager.getColor("key_text_color")
    private val mTypeface = FontManager.getTypeface("key_font")

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        val binding =
            SimpleItemRowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        val width = parent.dp(mSingleWidth)
        val height = parent.dp(mSingleHeight)
        val marginX = parent.dp(mStringMarginX).toInt()
        val bindings = mutableListOf<SimpleItemOneBinding>()
        for (i in 0 until columnSize) {
            val sub = SimpleItemOneBinding.inflate(LayoutInflater.from(parent.context), null, false)
            bindings.add(sub)
            val layoutParams = LinearLayout.LayoutParams(width, height)
            layoutParams.setMargins(marginX, 0, marginX, 0)
            binding.wrapper.addView(sub.root, layoutParams)
        }
        val holder = ViewHolder(binding, bindings)

        holder.simpleKeyTexts.forEachIndexed { index, textView ->
            holder.wrappers[index].tag = index
            textView.apply {
                textSize = mTextSize
                setTextColor(mTextColor)
                typeface = mTypeface
                gravity = Gravity.CENTER
                ellipsize = TextUtils.TruncateAt.MARQUEE
                background =
                    ColorManager.getDrawable(
                        "key_back_color",
                        "key_border_color",
                        dp(theme.generalStyle.keyBorder),
                        dp(theme.generalStyle.roundCorner),
                        cache = false,
                    )
            }
        }
        return holder
    }

    class ViewHolder(
        binding: SimpleItemRowBinding,
        views: List<SimpleItemOneBinding>,
    ) : RecyclerView.ViewHolder(binding.root) {
        val simpleKeyTexts = views.map { it.root.getChildAt(0) as TextView }
        val wrappers = views.map { it.root.apply { getChildAt(1).visibility = View.GONE } }
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int,
    ) {
        val bean = mBeansByRows[position]
        holder.simpleKeyTexts.forEachIndexed { index, textView ->
            textView.text = ""
            if (index < bean.size) {
                holder.wrappers[index].visibility = View.VISIBLE
                textView.text = bean[index].label
            } else {
                holder.wrappers[index].visibility = View.INVISIBLE
            }
            holder.wrappers[index]
                .setOnClickListener { view: View ->
                    if (view.tag != null) {
                        listener?.invoke(bean[index], position * columnSize + view.tag as Int)
                    }
                }
        }
    }

    /** 添加OnItemClickListener回调 */
    private var listener: (SimpleKeyBean.(Int) -> Unit)? = null

    fun setListener(listener: (SimpleKeyBean.(Int) -> Unit)?) {
        this.listener = listener
    }
}
